src/dune_lang/package_name.ml

```diff
@@ -15,8 +15,54 @@ include (
 
     let hint_valid = None
 
-    let of_string_opt s =
-      (* DUNE3 verify no dots or spaces *)
-      if s = "" then None else Some s
+    let of_string_opt s = if s = "" then None else Some s
   end) :
     Dune_util.Stringlike with type t := t)
+
+module Strict = struct
+  include Dune_util.Stringlike.Make (struct
+    type t = string
+
+    let module_ = "Package.Name.Strict"
+
+    let description = "package name"
+
+    let to_string s = s
+
+    let description_of_valid_string =
+      Some
+        (Pp.textf
+           "Package names can contain letters, '-', '_' and '+', and need to \
+            contain at least a letter.")
+
+    let of_string_opt s =
+      let state = ref (Ok false) in
+      String.iter s ~f:(fun c ->
+          match (!state, c) with
+          | Error (), _ -> ()
+          | _, ('a' .. 'z' | 'A' .. 'Z') -> state := Ok true
+          | _, ('0' .. '9' | '-' | '_' | '+') -> ()
+          | _ -> state := Error ());
```

Comment:
A little weird to keep iterating if you already found an error. Also, I'm not sure why you need three status. `type state = Valid | Invalid` would do and you can start with Invalid.

I also I see a couple of good alternatives:

1. Use a regular expression to make this code easier to understand
2. Use `OpamPackage.Name.of_string`. This will require catching `Failure _`.

Reply:
I started from the state machine from `OpamPackage.Name.of_string`. 3 states are necessary because there are 2 ways to be invalid: you encountered an invalid character (you can't recover from this), or you have not yet found a letter.

But I haven't realized I could use `re` in dune. I'll switch to that.

Reply:
Done. But I had to add `re` in `dune_lang/dune`. I'm not sure that's OK.

Reply:
It's not ideal. If the check is indeed so simple, you can just make it:

```
let valid = String.length x > 0 && String.forall ~f:is_valid_char in
```

But why can't you use the opam library directly? Opam is the source of truth for what is a valid package name. We should defer to it in such cases.